### PR TITLE
Support for "range" filter queries

### PR DESCRIPTION
**Pull request by FadhlanR**
Ticket: CS-6662

### CODE REVIEW

File: packages/runtime-common/indexer.ts [hunk 1]
```diff
@@ -339,6 +344,33 @@ export class Indexer {
     ]);
   }
 
+  private rangeCondition(filter: RangeFilter, on: CodeRef): CardExpression {
+    on = filter.on ?? on;
+    return every([
+      this.typeCondition(on),
+      ...flatten(
```
Comment (habdelra) [inline packages/runtime-common/indexer.ts]: Instead of using `flatten`, just splat `Object.entries`. Like we do in eqCondition and likeCondition above: `…Object.entries(  )`. It makes it easier to read the DSL that way

File: packages/runtime-common/indexer.ts [hunk 2]
```diff
@@ -339,6 +344,33 @@ export class Indexer {
...
+        Object.entries(filter.range).map(([path, range]) => {
+          let query = fieldQuery(path, on, false, 'filter');
+          let cardExpression: FieldArity[][] = [];
```
Comment (habdelra) [inline packages/runtime-common/indexer.ts]: I think this is better typed as `CardExpression`

File: packages/runtime-common/indexer.ts [hunk 3]
```diff
@@ -339,6 +344,33 @@ export class Indexer {
     ]);
   }
 
+  private rangeCondition(filter: RangeFilter, on: CodeRef): CardExpression {
+    on = filter.on ?? on;
+    return every([
+      this.typeCondition(on),
+      ...flatten(
+        Object.entries(filter.range).map(([path, range]) => {
+          let query = fieldQuery(path, on, false, 'filter');
```
Comment (habdelra) [inline packages/runtime-common/indexer.ts]: Let’s follow the same patten as the likeCondition and eqCondition methods and split the inside of this Object.entries into a different method: fieldRangeFilter. Id like to have this line up with those other two filters closely in their structure so that the code reads consistently.

File: packages/runtime-common/tests/query-test.ts [hunk 4]
```diff
@@ -1619,6 +1619,424 @@ const tests = Object.freeze({
       ]);
     }
   },
+
```
Comment (habdelra) [inline packages/runtime-common/tests/query-test.ts:1622]: Let’s add a test that can filter range thru a plural field:
Like `range: { friends.age: { gt: 3 } }` where friends is a linksToMany.

Comment (habdelra) [reply packages/runtime-common/tests/query-test.ts:1622]: Let’s also add a test where the range field we are testing itself is a containsMany. Maybe like `lotteryNumbers` containsMany field, and you want to find all the people who choose lottery numbers above 50 (you have to add a new field to the host and server test cards)

Comment (FadhlanR) [reply packages/runtime-common/tests/query-test.ts:1622]: Added the tests [650053e](https://github.com/cardstack/boxel/pull/1194/commits/650053ed178b2c2bca48f9c125b81fadebfeb010)

File: packages/runtime-common/indexer.ts (hunk 3 above)
Comment (FadhlanR) [reply packages/runtime-common/indexer.ts]: Updated [650053e](https://github.com/cardstack/boxel/pull/1194/commits/650053ed178b2c2bca48f9c125b81fadebfeb010)

File: packages/runtime-common/indexer.ts [hunk 5]
```diff
@@ -399,6 +415,30 @@ export class Indexer {
...
 
+  private fieldRangeFilter(
+    key: string,
+    filterValue: RangeFilterValue,
+    onRef: CodeRef,
+  ): CardExpression {
+    let query = fieldQuery(key, onRef, false, 'filter');
+    let cardExpressions: CardExpression[] = [];
+    Object.entries(filterValue).forEach(([operator, value]) => {
+      if (value != null) {
```
Comment (habdelra) [inline packages/runtime-common/indexer.ts]: change to:
```suggestion
if (value == null) {
  throw new Error(`'null' is not a permitted value in a 'range' filter`);
}
```

Comment (habdelra) [reply packages/runtime-common/indexer.ts]: also add a test that we throw if the value in a range filter is null

**Comment by habdelra** [comment]
looks really good just a couple more things
